app/src/org/commcare/dalvik/activities/HomeActivityUIController.java

```diff
@@ -0,0 +1,391 @@
+        }
+        adapter.setOnClickListenerForButton(R.layout.home_incompleteforms_button, getIncompleteButtonListener());
+    }
+
+    private void setupLogoutButton() {
+        logoutButton = adapter.getButton(R.layout.home_disconnect_button);
+        if (logoutButton != null) {
+            logoutButton.setText(Localization.get("home.logout"));
+            logoutButton.setNotificationText(activity.getActivityTitle());
+            adapter.notifyDataSetChanged();
+        }
+        adapter.setOnClickListenerForButton(R.layout.home_disconnect_button, getLogoutButtonListener());
+    }
+
+    private void setupViewOldFormsButton() {
+        SquareButtonWithNotification viewOldForms = adapter.getButton(R.layout.home_savedforms_button);
+        if (viewOldForms != null) {
+            viewOldForms.setText(Localization.get("home.forms.saved"));
+        }
+        adapter.setOnClickListenerForButton(R.layout.home_savedforms_button, getViewOldFormsListener());
+    }
+
+    private void setupSyncButton() {
+        syncButton = adapter.getButton(R.layout.home_sync_button);
+        if (syncButton != null) {
+            setSyncButtonText(CommCareApplication._().getSyncDisplayParameters(), null);
+        }
+        adapter.setOnClickListenerForButton(R.layout.home_sync_button, getSyncButtonListener());
+    }
+
+    // endregion
+
+
+    // region - generators for the listeners for each button
+
+    private View.OnClickListener getViewOldFormsListener() {
+        return new View.OnClickListener() {
+            public void onClick(View v) {
+                activity.goToFormArchive(false);
+            }
+        };
+    }
+
+    private View.OnClickListener getSyncButtonListener() {
+        return new View.OnClickListener() {
+            public void onClick(View v) {
+                activity.attemptSync();
+            }
+        };
+    }
+
+    private View.OnClickListener getStartButtonListener() {
+        return new View.OnClickListener() {
+            public void onClick(View v) {
+                activity.enterRootModule();
+            }
+        };
+    }
+
+    private View.OnClickListener getIncompleteButtonListener() {
+        return new View.OnClickListener() {
+            public void onClick(View v) {
+                activity.goToFormArchive(true);
+            }
+        };
+    }
+
+    private View.OnClickListener getLogoutButtonListener() {
+        return new View.OnClickListener() {
+            public void onClick(View v) {
+                CommCareApplication._().closeUserSession();
+                activity.returnToLogin();
+            }
+        };
+    }
+
+    // endregion
+
+
+    // region - text setters for all buttons
+
+    private void setSyncButtonText(Pair<Long, int[]> syncDetails, String syncTextKey) {
+        if (syncTextKey == null) {
+            syncTextKey = activity.isDemoUser() ? "home.sync.demo" : "home.sync";
```

Comment:
Haha, I do have a perhaps illogical hatred of ternary operators! Not gonna worrying about it here because it is copied, not new, code :)